portal/app/farm_updater/egauge_api.py

```diff
@@ -0,0 +1,128 @@
+import time
+import json
+
+class EgaugeInterface():
+
+    def __init__(self, url=None, username=None, password=None, t_sample=5):
+        # Initializing credentials
+        self.url = url
+        self.username = username
+        self.password = password
+
+        # Initializing parameters
+        self.t_sample = t_sample
+        self.keys = ['L1 Voltage', 'L2 Voltage', 'Power Circuit 1', 'Power Circuit 1*', 'Power Circuit 2',
+                     'Power Circuit 2*', 'Power Circuit 1 neutral', 'Shed Power', 'Control Fan Power',
+                     'Control Fan Power*', 'ts']
+
+    # Function to get and format e-gauge data
+    def get_egauge_data(self, request):
+        power_values = dict.fromkeys(self.keys, None)
+        root = ET.fromstring(request.text)
+        timestamp = root.findtext("ts")
+
+        if timestamp != None:
+            for r in root.findall('r'):
+                for child in r:
+                    if r.get('n') == 'L1 Voltage':
+                        power_values['L1 Voltage'] = (int(child.text))
+                    elif r.get('n') == 'L2 Voltage':
+                        power_values['L2 Voltage'] = (int(child.text))
+                    elif r.get('n') == 'Power Circuit 1':
+                        power_values['Power Circuit 1'] = (int(child.text))
+                    elif r.get('n') == 'Power Circuit 1*':
+                        power_values['Power Circuit 1*'] = (int(child.text))
+                    elif r.get('n') == 'Power Circuit 2':
+                        power_values['Power Circuit 2'] = (int(child.text))
+                    elif r.get('n') == 'Power Circuit 2*':
+                        power_values['Power Circuit 2*'] = (int(child.text))
+                    elif r.get('n') == 'Power Circuit 1 neutral':
+                        power_values['Power Circuit 1 neutral'] = (int(child.text))
+                    elif r.get('n') == 'Shed Power':
+                        power_values['Shed Power'] = (int(child.text))
+                    elif r.get('n') == 'Control Fan Power':
+                        power_values['Control Fan Power'] = (int(child.text))
+                    elif r.get('n') == 'Control Fan Power*':
+                        power_values['Control Fan Power*'] = (int(child.text))
+
+            power_values['ts'] = int(timestamp)
+
+        else:
+            print('No values from request')
+
+        return power_values
+
+    # Function to process data from e-gauge and convert to useful power values
+    def processing_egauge_data(self):
+        power_values = dict.fromkeys(self.keys, None)
+        try:
+            resp = requests.get(self.url, auth=HTTPDigestAuth(self.username, self.password))
+            resp.raise_for_status()
+            data_ini = self.get_egauge_data(resp)
+
+        except requests.exceptions.HTTPError as err:
+            print(err)
+            return json.dumps(power_values)
+
+        time.sleep(self.t_sample)
```

Comment:
what is this sleep for? This is really weird

Reply:
need this bc the way egauge registers accumulate values. I need to do two consecutive requests (I could remove this sleep but not sure how egauge server will respond - they recommend having a delay) to get the registers and divide the difference of these registers. I could use the previous time value (in the original format) and store it - just not sure if there's an efficient way to do this

Reply:
Hummmmm
Does it make sense to reduce the interval of the updater instead? Then on every new entry you just make this same calculation against the previous DB entry and then update our records accordingly? (instead of comparing two entries from the network)

Reply:
I was thinking about this initially but could not test on jupyter. Will need to learn how to do it in django but this should work.

Reply:
Will let's do it before we merge this into master - it should simplify this logic. I can show you an example.

Reply:
done